### PR TITLE
feat(core): add factory-provider circular dependency check

### LINKED ISSUE #11093: Circular dependency between two factory providers (add error message & throw)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

```typescript
export const dynamicProvider1 = {
  provide: 'PROVIDER1',
  useFactory: (provider2: any) => {
    setTimeout(() => {
      console.log('from provider 1', provider2);
    }, 1);
    return {
      name: 'provider 1',
    };
  },
  inject: ['PROVIDER2'],
};

export const dynamicProvider2 = {
  provide: 'PROVIDER2',
  useFactory: (provider1: any) => {
    setTimeout(() => {
      console.log('from provider 2', provider1);
    }, 1);
    return {
      name: 'provider 2',
    };
  },
  inject: ['PROVIDER1'],
};
```

No errors, nothing output.

### Minimum reproduction code

https://github.com/lybrus/dynamic-providers-circular-dependency

### Steps to reproduce

1. npm i
2. npm run start:dev
3. The application hangs on `LOG [NestFactory] Starting Nest application...`

### Expected behavior

Error

or

Output in some sequence, but I have no idea how to use forwardRef in this case.

`
from provider 2 { name: "provider 1" }
from provider 1 { name: "provider 2" }
`

### Package

- [X] I don't know. Or some 3rd-party package
- [ ] <code>@nestjs/common</code>
- [ ] <code>@nestjs/core</code>
- [ ] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [ ] Other (see below)

### Other package

_No response_

### NestJS version

_No response_

### Packages versions

```json

```

### Node.js version

_No response_

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

### Other

It would be better to have an error message.

### PR DESCRIPTION
**Pull request by Flusinerd**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #11093 

Currently the nest application won't start when there is a circular dependency between factory providers

## What is the new behavior?

A descriptive error will be thrown when a circular dependency is detected.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: integration/injector/e2e/circular-factory-providers.spec.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      CircularDependencyFactoryProviderException,
+    );
+  });
+
+  it('should log no Error when no Circular Dependency between Factory Providers is detected', async () => {
+    const builder = Test.createTestingModule({
+      imports: [NonCircularFactoryProvidersModule],
+    });
+
+    // Assert that the Error is thrown
```
Comment (micalevisk) [inline integration/injector/e2e/circular-factory-providers.spec.ts:33]:
```suggestion
```

File: packages/core/injector/container.ts [hunk 2]
```diff
@@ -172,9 +183,68 @@ export class NestContainer {
     if (!moduleRef) {
       throw new UnknownModuleException();
     }
+    // Detect circular dependencies between existing providers and the new one
+    if (isFactoryProvider(provider)) {
```
Comment (kamilmysliwiec) [inline packages/core/injector/container.ts:187]: Are we sure that we want to run `detectCircularDependencyBetweenFactoryProviders` for every single factory provider?

AFAIR we should be able to fix this issue by tweaking Injector's implementation. The container class shouldn't be responsible for this

**Review by kamilmysliwiec — CHANGES_REQUESTED** [review]

**Comment by Flusinerd** [comment]
Hi @kamilmysliwiec 
Sorry I did not have much time lately but want to get this solved now.

I did some further research and can think of this solution:

We could make use of the ``InstanceMetadataStore.dependencies`` array to check for circular dependencies inside the injector:

For example, these provider construct:
```ts
export const dynamicProvider1 = {
  provide: 'PROVIDER1',
  useFactory: (provider2: any) => {
    return {
      name: 'provider 1',
    };
  },
  inject: ['PROVIDER2'],
};

export const dynamicProvider2 = {
  provide: 'PROVIDER2',
  useFactory: (provider1: any) => {
    return {
      name: 'provider 2',
    };
  },
  inject: ['PROVIDER1'],
};
```
would result in this while resolving ``PROVIDER1`` for ``PROVIDER2``

``InstanceWrapper(PROVIDER1).instance_metadata``:
```
dependencies: [{
  name: 'PROVIDER2' // Everything is fine here
  instance_metadata: {
    dependencies: [{
      name: 'PROVIDER1'// HERE the circular dependencie occurs, because we encounter PROVIDER1 while resolving PROVIDER1
    }]
}];
```

Would this be a suitable solution?

This check could for example occur here:
https://github.com/nestjs/nest/blob/883d0fb8a874b359c54dd541cde206c9c23d790e/packages/core/injector/injector.ts#L514

**Comment by Flusinerd** [comment]
This would be my proposed solution:
```ts
public async resolveComponentInstance<T>(
    moduleRef: Module,
    token: InjectionToken,
    dependencyContext: InjectorDependencyContext,
    wrapper: InstanceWrapper<T>,
    contextId = STATIC_CONTEXT,
    inquirer?: InstanceWrapper,
    keyOrIndex?: symbol | string | number,
  ): Promise<InstanceWrapper> {
    this.printResolvingDependenciesLog(token, inquirer);
    this.printLookingForProviderLog(token, moduleRef);
    const providers = moduleRef.providers;
    const instanceWrapper = await this.lookupComponent(
      providers,
      moduleRef,
      { ...dependencyContext, name: token },
      wrapper,
      contextId,
      inquirer,
      keyOrIndex,
    );
   
    // Start of changes
    if (!isNil(instanceWrapper.inject)) {
      const encounteredDependencies = new Set<InjectionToken>();
      const checkForCircularDependency = (instance: InstanceWrapper) => {
        const dependencies = instance[INSTANCE_METADATA_SYMBOL].dependencies;
        dependencies.forEach(dep => {
          if (encounteredDependencies.has(dep.token)) {
            const injectionPath = [...encounteredDependencies, dep.token].map(
              token => token.toString(),
            );
            throw new CircularDependencyFactoryProviderException(injectionPath);
          }

          if (!isNil(dep.inject)) {
            encounteredDependencies.add(dep.token);
            checkForCircularDependency(dep);
          }
        });
      };

      checkForCircularDependency(instanceWrapper);
    }
    // End of changes

    return this.resolveComponentHost(
      moduleRef,
      instanceWrapper,
      contextId,
      inquirer,
    );
  }
```

This is basically using the Instancewrapper metadata as a dependency tree and a depth first search to detect the circular dependency utilizing the injection token. So this also works for: ``PROVIDER1 -> PROVIDER2 -> PROVIDER3 -> PROVIDER1``

**Pull request closed without merging**